In [154]:
import os
os.environ['kaggle_config_dir'] = '/content/drive/MyDrive/Colab Notebooks'

In [155]:
pwd

'/content/drive/MyDrive/Colab Notebooks'

In [156]:
%cd /content/drive/MyDrive/Colab Notebooks

/content/drive/MyDrive/Colab Notebooks


In [157]:
pwd

'/content/drive/MyDrive/Colab Notebooks'

In [158]:
!kaggle datasets download -d ahmedmohamed2003/cafe-sales-dirty-data-for-cleaning-training

Dataset URL: https://www.kaggle.com/datasets/ahmedmohamed2003/cafe-sales-dirty-data-for-cleaning-training
License(s): CC-BY-SA-4.0
100% 111k/111k [00:00<00:00, 393kB/s]
100% 111k/111k [00:00<00:00, 388kB/s]


In [6]:
!unzip \*.zip && rm *.zip

Archive:  cafe-sales-dirty-data-for-cleaning-training.zip
replace dirty_cafe_sales.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: n


In [159]:
import pandas as pd

df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/dirty_cafe_sales.csv')

df1 = df.copy()
df2 = df.copy()


In [175]:
import numpy as np

In [160]:
df.head()

Transaction ID    Item Quantity Price Per Unit Total Spent  Payment Method  \
0    TXN_1961373  Coffee        2            2.0         4.0     Credit Card   
1    TXN_4977031    Cake        4            3.0        12.0            Cash   
2    TXN_4271903  Cookie        4            1.0       ERROR     Credit Card   
3    TXN_7034554   Salad        2            5.0        10.0         UNKNOWN   
4    TXN_3160411  Coffee        2            2.0         4.0  Digital Wallet   

   Location Transaction Date  
0  Takeaway       2023-09-08  
1  In-store       2023-05-16  
2  In-store       2023-07-19  
3   UNKNOWN       2023-04-27  
4  In-store       2023-06-11

In [161]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Transaction ID    10000 non-null  object
 1   Item              9667 non-null   object
 2   Quantity          9862 non-null   object
 3   Price Per Unit    9821 non-null   object
 4   Total Spent       9827 non-null   object
 5   Payment Method    7421 non-null   object
 6   Location          6735 non-null   object
 7   Transaction Date  9841 non-null   object
dtypes: object(8)
memory usage: 625.1+ KB


In [162]:
df.shape

(10000, 8)

## Checking for duplicates

In [164]:
df.duplicated().sum()

0

There were no duplicates identiifed

## Finding number of empty values for each variable

In [35]:
df.isnull().sum()

Transaction ID         0
Item                 333
Quantity             138
Price Per Unit       179
Total Spent          173
Payment Method      2579
Location            3265
Transaction Date     159
dtype: int64

In [165]:
df['Transaction Date'].isnull().value_counts(dropna= False)

Transaction Date
False    9841
True      159
Name: count, dtype: int64

Above result shows there are 159 rows are empty. Trying to check if there are any other values apart from date in reamining rows

In [43]:
df['Transaction Date'].value_counts()

Transaction Date
UNKNOWN       159
ERROR         142
2023-06-16     40
2023-02-06     40
2023-07-21     39
             ... 
2023-04-27     15
2023-09-24     15
2023-07-22     14
2023-03-11     14
2023-02-17     14
Name: count, Length: 367, dtype: int64

Performing value count operation show there are two types of values 'UNKNOWN' and 'ERROR'

In [166]:
df['Transaction Date'] = pd.to_datetime(df['Transaction Date'], errors= "coerce")

In [167]:
df['Transaction Date'].isnull().value_counts(dropna= False)

Transaction Date
False    9540
True      460
Name: count, dtype: int64

Above result shows updated null values present in the column compared to previous result for checking null values in Transaction Date column

## Performing similar check for other columns to see if they have 'ERROR' and 'UNKNOWN' values in the rows apart from null values of respective variables

In [168]:
df['Location'].value_counts()

Location
Takeaway    3022
In-store    3017
ERROR        358
UNKNOWN      338
Name: count, dtype: int64

In [169]:
df['Total Spent'].value_counts()

Total Spent
6.0        979
12.0       939
3.0        930
4.0        923
20.0       746
15.0       734
8.0        677
10.0       524
2.0        497
9.0        479
5.0        468
16.0       444
25.0       259
7.5        237
1.0        232
4.5        225
1.5        205
UNKNOWN    165
ERROR      164
Name: count, dtype: int64

In [170]:
df['Payment Method'].value_counts()

Payment Method
Digital Wallet    2291
Credit Card       2273
Cash              2258
ERROR              306
UNKNOWN            293
Name: count, dtype: int64

In [171]:
df['Item'].value_counts()

Item
Juice       1171
Coffee      1165
Salad       1148
Cake        1139
Sandwich    1131
Smoothie    1096
Cookie      1092
Tea         1089
UNKNOWN      344
ERROR        292
Name: count, dtype: int64

In [172]:
df['Price Per Unit'].value_counts()

Price Per Unit
3.0        2429
4.0        2331
2.0        1227
5.0        1204
1.0        1143
1.5        1133
ERROR       190
UNKNOWN     164
Name: count, dtype: int64

In [174]:
df['Quantity'].value_counts()

Quantity
5          2013
2          1974
4          1863
3          1849
1          1822
UNKNOWN     171
ERROR       170
Name: count, dtype: int64

After seeing the values in all columns there are some rows with Error, Unknown values apart from blank rows.

So, they need to be replaced with np.nan in order to show them as blank rows

In [177]:
values = ["ERROR", "UNKNOWN"]

In [178]:
df.replace(values, np.nan, inplace= True)

In [183]:
df[["Quantity","Price Per Unit", "Total Spent"]] = df[["Quantity","Price Per Unit", "Total Spent"]].apply(pd.to_numeric, errors= "coerce")


In [184]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    10000 non-null  object        
 1   Item              9031 non-null   object        
 2   Quantity          9521 non-null   float64       
 3   Price Per Unit    9467 non-null   float64       
 4   Total Spent       9498 non-null   float64       
 5   Payment Method    6822 non-null   object        
 6   Location          6039 non-null   object        
 7   Transaction Date  9540 non-null   datetime64[ns]
dtypes: datetime64[ns](1), float64(3), object(4)
memory usage: 625.1+ KB


Numerical columns are changed to respective data type and rows with unknown and error values are replaced with np.nan values

In [186]:
df[df['Price Per Unit'].isnull()]

Transaction ID      Item  Quantity  Price Per Unit  Total Spent  \
56      TXN_3578141      Cake       5.0             NaN         15.0   
65      TXN_4987129  Sandwich       3.0             NaN          NaN   
68      TXN_8427104     Salad       2.0             NaN         10.0   
85      TXN_8035512       Tea       3.0             NaN          4.5   
104     TXN_7447872     Juice       2.0             NaN          6.0   
...             ...       ...       ...             ...          ...   
9924    TXN_5981429     Juice       2.0             NaN          6.0   
9926    TXN_2464706      Cake       4.0             NaN         12.0   
9961    TXN_2153100       Tea       2.0             NaN          3.0   
9996    TXN_9659401       NaN       3.0             NaN          3.0   
9998    TXN_7695629    Cookie       3.0             NaN          3.0   

      Payment Method  Location Transaction Date  
56               NaN  Takeaway       2023-06-27  
65               NaN  In-store       2023-10-20  
68               NaN  In-store       2023-10-27  
85              Cash       NaN       2023-10-29  
104              NaN       NaN              NaT  
...              ...       ...              ...  
9924  Digital Wallet       NaN       2023-12-24  
9926  Digital Wallet  Takeaway       2023-11-09  
9961            Cash       NaN       2023-12-29  
9996  Digital Wallet       NaN       2023-06-02  
9998  Digital Wallet       NaN       2023-12-02  

[533 rows x 8 columns]

### Assumption is that the price per unit item does not change

Viewing the above table it is clear we can extract the price of items and map corresponing price to replace NaN values in 'Price Per Unit' column

In [110]:
unique_items = [item for item in df['Item'].unique() if item is not np.nan]
unique_items

['Coffee', 'Cake', 'Cookie', 'Salad', 'Smoothie', 'Sandwich', 'Juice', 'Tea']

In [112]:
item_price_dict = {}

for item in unique_items:
  item_price_dict[item] = df[df['Item']== item]['Price Per Unit'].value_counts().index

item_price_dict

{'Coffee': Index([2.0], dtype='float64', name='Price Per Unit'),
 'Cake': Index([3.0], dtype='float64', name='Price Per Unit'),
 'Cookie': Index([1.0], dtype='float64', name='Price Per Unit'),
 'Salad': Index([5.0], dtype='float64', name='Price Per Unit'),
 'Smoothie': Index([4.0], dtype='float64', name='Price Per Unit'),
 'Sandwich': Index([4.0], dtype='float64', name='Price Per Unit'),
 'Juice': Index([3.0], dtype='float64', name='Price Per Unit'),
 'Tea': Index([1.5], dtype='float64', name='Price Per Unit')}

In [113]:
# cleaning up item_price_dict

for key in item_price_dict:
  item_price_dict[key] = item_price_dict[key][0]

item_price_dict

{'Coffee': 2.0,
 'Cake': 3.0,
 'Cookie': 1.0,
 'Salad': 5.0,
 'Smoothie': 4.0,
 'Sandwich': 4.0,
 'Juice': 3.0,
 'Tea': 1.5}

Unique value of items are extracted from the column and respective prices are saved into a dictonary of key value pairs for the purpose of mapping with 'Price Per Unit' column

In [187]:
df['Price Per Unit'] = df['Price Per Unit'].fillna(df['Item'].map(item_price_dict))

In [188]:
df['Price Per Unit'].isna().sum()

54

Assuming same logic as price of item doesn't for an item and vice versa

#### Reversing the key value pairs in dictionary to get the item value with respect to price.


In [189]:
price_item = { price:item for item,price in item_price_dict.items()}
price_item

{2.0: 'Coffee',
 3.0: 'Juice',
 1.0: 'Cookie',
 5.0: 'Salad',
 4.0: 'Sandwich',
 1.5: 'Tea'}

In [190]:
df['Item'] = df['Item'].fillna(df['Price Per Unit'].map(price_item))

In [191]:
df['Item'].isna().value_counts()

Item
False    9946
True       54
Name: count, dtype: int64

In [192]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    10000 non-null  object        
 1   Item              9946 non-null   object        
 2   Quantity          9521 non-null   float64       
 3   Price Per Unit    9946 non-null   float64       
 4   Total Spent       9498 non-null   float64       
 5   Payment Method    6822 non-null   object        
 6   Location          6039 non-null   object        
 7   Transaction Date  9540 non-null   datetime64[ns]
dtypes: datetime64[ns](1), float64(3), object(4)
memory usage: 625.1+ KB


In [193]:
df['Total Spent'].isna().value_counts(dropna= False)

Total Spent
False    9498
True      502
Name: count, dtype: int64

In [194]:
df['Total Spent'] = df['Total Spent'].fillna(df['Quantity']*df['Price Per Unit'])

In [195]:
df['Total Spent'].isna().sum()

23

In [196]:
df[df['Total Spent'].isna()]

Transaction ID      Item  Quantity  Price Per Unit  Total Spent  \
236     TXN_8562645     Salad       NaN             5.0          NaN   
278     TXN_3229409     Juice       NaN             3.0          NaN   
641     TXN_2962976     Juice       NaN             3.0          NaN   
738     TXN_8696094  Sandwich       NaN             4.0          NaN   
1761    TXN_3611851       NaN       4.0             NaN          NaN   
2289    TXN_7524977       NaN       4.0             NaN          NaN   
2796    TXN_9188692      Cake       NaN             3.0          NaN   
3203    TXN_4565754  Smoothie       NaN             4.0          NaN   
3224    TXN_6297232    Coffee       NaN             2.0          NaN   
3401    TXN_3251829       Tea       NaN             1.5          NaN   
4152    TXN_9646000       NaN       2.0             NaN          NaN   
4257    TXN_6470865    Coffee       NaN             2.0          NaN   
5841    TXN_5884081    Cookie       NaN             1.0          NaN   
7029    TXN_4628338    Coffee       NaN             2.0          NaN   
7297    TXN_9944500  Smoothie       NaN             4.0          NaN   
8021    TXN_2428781     Salad       NaN             5.0          NaN   
8443    TXN_2023651  Sandwich       NaN             4.0          NaN   
8465    TXN_9669616    Coffee       NaN             2.0          NaN   
8479    TXN_1547245  Sandwich       NaN             4.0          NaN   
8574    TXN_2546684     Juice       NaN             3.0          NaN   
8732    TXN_4550558    Cookie       NaN             1.0          NaN   
9590    TXN_9924732  Sandwich       NaN             4.0          NaN   
9869    TXN_1975184    Coffee       NaN             2.0          NaN   

      Payment Method  Location Transaction Date  
236              NaN  In-store       2023-05-18  
278             Cash  Takeaway       2023-04-15  
641              NaN       NaN       2023-03-17  
738              NaN  Takeaway       2023-05-14  
1761     Credit Card       NaN       2023-02-09  
2289             NaN       NaN       2023-12-09  
2796     Credit Card       NaN       2023-12-01  
3203  Digital Wallet  Takeaway       2023-10-06  
3224             NaN       NaN       2023-04-07  
3401  Digital Wallet  In-store       2023-07-25  
4152             NaN  In-store       2023-12-14  
4257  Digital Wallet  Takeaway       2023-09-18  
5841  Digital Wallet  In-store       2023-07-05  
7029            Cash       NaN       2023-12-25  
7297            Cash  In-store       2023-01-03  
8021             NaN  In-store       2023-05-09  
8443            Cash  In-store       2023-05-25  
8465             NaN       NaN       2023-06-03  
8479             NaN  Takeaway       2023-09-11  
8574  Digital Wallet  Takeaway       2023-04-08  
8732     Credit Card  In-store       2023-08-04  
9590     Credit Card  In-store       2023-01-18  
9869  Digital Wallet       NaN       2023-01-15

In [197]:
df[df['Quantity'].isna()]

Transaction ID      Item  Quantity  Price Per Unit  Total Spent  \
20      TXN_3522028  Smoothie       NaN             4.0         20.0   
55      TXN_5522862    Cookie       NaN             1.0          2.0   
57      TXN_2080895      Cake       NaN             3.0          3.0   
66      TXN_8501819     Juice       NaN             3.0          6.0   
117     TXN_2148617     Juice       NaN             3.0          9.0   
...             ...       ...       ...             ...          ...   
9932    TXN_8502079       Tea       NaN             1.5          3.0   
9935    TXN_9778251       Tea       NaN             1.5          6.0   
9944    TXN_7495283      Cake       NaN             3.0         15.0   
9957    TXN_6487003    Coffee       NaN             2.0          8.0   
9984    TXN_3142496  Smoothie       NaN             4.0          4.0   

      Payment Method  Location Transaction Date  
20              Cash  In-store       2023-04-04  
55       Credit Card  Takeaway       2023-03-19  
57    Digital Wallet  In-store       2023-04-19  
66              Cash       NaN       2023-03-30  
117   Digital Wallet       NaN       2023-01-10  
...              ...       ...              ...  
9932            Cash       NaN       2023-04-20  
9935             NaN  Takeaway       2023-11-09  
9944     Credit Card  Takeaway       2023-04-14  
9957     Credit Card  Takeaway       2023-11-15  
9984            Cash  Takeaway       2023-07-27  

[479 rows x 8 columns]